{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DNA Feature Viewer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the `dna_features_viewer` library here: https://github.com/Edinburgh-Genome-Foundry/DnaFeaturesViewer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Preprint - https://www.biorxiv.org/content/10.1101/2020.01.09.900589v1.full.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# pip install dna_features_viewer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# pip install moviepy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# pip install bcbio-gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dna_features_viewer import GraphicFeature, GraphicRecord\n",
    "features=[\n",
    "    GraphicFeature(start=0, end=20, strand=+1, color=\"#ffd700\",\n",
    "                   label=\"Small feature\"),\n",
    "    GraphicFeature(start=20, end=500, strand=+1, color=\"#ffcccc\",\n",
    "                   label=\"Gene 1 with a very long name\"),\n",
    "    GraphicFeature(start=400, end=700, strand=-1, color=\"#cffccc\",\n",
    "                   label=\"Gene 2\"),\n",
    "    GraphicFeature(start=600, end=900, strand=+1, color=\"#ccccff\",\n",
    "                   label=\"Gene 3\")\n",
    "]\n",
    "record = GraphicRecord(sequence_length=1000, features=features)\n",
    "record.plot(figure_width=5);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load overview_and_detail.py\n",
    "\"\"\"Generate a 2-plot figure with full sequence on the left,\n",
    "detail of a sub-segment on the right.\"\"\"\n",
    "\n",
    "from dna_features_viewer import GraphicFeature, GraphicRecord\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "record = GraphicRecord(sequence=250 * \"ATGC\", features=[\n",
    "    GraphicFeature(start=10, end=20, strand=+1, color=\"#ffd700\",\n",
    "                   label=\"Small feature\"),\n",
    "    GraphicFeature(start=20, end=500, strand=+1, color=\"#ffcccc\",\n",
    "                   label=\"Gene 1 with a very long name\"),\n",
    "    GraphicFeature(start=400, end=700, strand=-1, color=\"#cffccc\",\n",
    "                   label=\"Gene 2\"),\n",
    "    GraphicFeature(start=600, end=900, strand=+1, color=\"#ccccff\",\n",
    "                   label=\"Gene 3\")\n",
    "])\n",
    "zoom_start, zoom_end = 398, 428  # coordinates of the \"detail\"\n",
    "cropped_record = record.crop((zoom_start, zoom_end))\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 2))\n",
    "\n",
    "# PLOT THE WHOLE SEQUENCE\n",
    "\n",
    "ax1.set_title(\"Whole sequence\", loc='left', weight='bold')\n",
    "record.plot(ax=ax1)\n",
    "ax1.fill_between((zoom_start, zoom_end), +1000, -1000, alpha=0.15)\n",
    "\n",
    "# PLOT THE SEQUENCE DETAILS\n",
    "\n",
    "\n",
    "cropped_record.plot_translation(ax=ax2, location=(408, 423),\n",
    "                                fontdict={'weight': 'bold'})\n",
    "cropped_record.plot(ax=ax2, plot_sequence=True)\n",
    "ax2.set_title(\"Sequence detail\", loc='left', weight='bold')\n",
    "\n",
    "fig.savefig('overview_and_detail.png', bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load from_genbank.py \n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "\n",
    "graphic_record = BiopythonTranslator().translate_record(\"example_sequence.gb\")\n",
    "ax, _ = graphic_record.plot(figure_width=10, strand_in_label_threshold=7)\n",
    "ax.figure.tight_layout()\n",
    "ax.figure.savefig(\"from_genbank.png\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load example_with_inverted_x_axis.py\n",
    "\"\"\"This example shows how you can very easily flip a plot horizontally if you\n",
    "need, using Matplotlib's ax.set_xlim() method.\"\"\"\n",
    "\n",
    "from dna_features_viewer import BiopythonTranslator, load_record\n",
    "\n",
    "ax = BiopythonTranslator.quick_class_plot(\"example_sequence.gb\")\n",
    "x1, x2 = ax.get_xlim()\n",
    "ax.set_xlim(x2, x1)\n",
    "ax.figure.tight_layout()\n",
    "ax.figure.savefig(\"example_with_inverted_x_axis.svg\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load sequence_and_translation.py\n",
    "from dna_features_viewer import GraphicFeature, GraphicRecord\n",
    "\n",
    "record = GraphicRecord(sequence=\"ATGCATGCATGCATGCATGCATGCATGC\", features=[\n",
    "    GraphicFeature(start=5, end=10, strand=+1, color='#ffcccc'),\n",
    "    GraphicFeature(start=8, end=15, strand=+1, color='#ccccff')\n",
    "])\n",
    "\n",
    "ax, _ = record.plot(figure_width=6)\n",
    "record.plot_sequence(ax, guides_intensity=0.2)\n",
    "fontdict = {'weight': 'bold'}\n",
    "record.plot_translation(ax, (8, 23), fontdict=fontdict, guides_intensity=0.8)\n",
    "ax.figure.savefig('sequence_and_translation.png', bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load custom_biopython_translator.py\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "\n",
    "class MyCustomTranslator(BiopythonTranslator):\n",
    "    \"\"\"Custom translator implementing the following theme:\n",
    "\n",
    "    - Color terminators in green, CDS in blue, all other features in gold.\n",
    "    - Do not display features that are restriction sites unless they are BamHI\n",
    "    - Do not display labels for restriction sites.\n",
    "    - For CDS labels just write \"CDS here\" instead of the name of the gene.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def compute_feature_color(self, feature):\n",
    "        if feature.type == \"CDS\":\n",
    "            return \"blue\"\n",
    "        elif feature.type == \"terminator\":\n",
    "            return \"green\"\n",
    "        else:\n",
    "            return \"gold\"\n",
    "\n",
    "    def compute_feature_label(self, feature):\n",
    "        if feature.type == 'restriction_site':\n",
    "            return None\n",
    "        elif feature.type == \"CDS\":\n",
    "            return \"CDS here\"\n",
    "        else:\n",
    "            return BiopythonTranslator.compute_feature_label(self, feature)\n",
    "\n",
    "    def compute_filtered_features(self, features):\n",
    "        \"\"\"Do not display promoters. Just because.\"\"\"\n",
    "        return [\n",
    "            feature for feature in features\n",
    "            if (feature.type != \"restriction_site\")\n",
    "            or (\"BamHI\" in str(feature.qualifiers.get(\"label\", '')))\n",
    "        ]\n",
    "\n",
    "\n",
    "graphic_record = MyCustomTranslator().translate_record(\"example_sequence.gb\")\n",
    "ax, _ = graphic_record.plot(figure_width=15)\n",
    "ax.figure.tight_layout()\n",
    "ax.figure.savefig(\"custom_biopython_translator.svg\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load gff_record_from_the_web.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load graphic_record_defined_by_hand.py\n",
    "\"\"\"Simple example where a few features are defined \"by hand\" and are displayed\n",
    "and exported as PNG, first with a linear view, then with a circular\n",
    "view.\n",
    "\"\"\"\n",
    "\n",
    "from dna_features_viewer import (\n",
    "    GraphicFeature,\n",
    "    GraphicRecord,\n",
    "    CircularGraphicRecord,\n",
    ")\n",
    "\n",
    "features = [\n",
    "    GraphicFeature(\n",
    "        start=5, end=20, strand=+1, color=\"#ffd700\", label=\"Small feature\"\n",
    "    ),\n",
    "    GraphicFeature(\n",
    "        start=20,\n",
    "        end=500,\n",
    "        strand=+1,\n",
    "        color=\"#ffcccc\",\n",
    "        label=\"Gene 1 with a very long name\",\n",
    "    ),\n",
    "    GraphicFeature(\n",
    "        start=400, end=700, strand=-1, color=\"#cffccc\", label=\"Gene 2\"\n",
    "    ),\n",
    "    GraphicFeature(\n",
    "        start=600, end=900, strand=+1, color=\"#ccccff\", label=\"Gene 3\"\n",
    "    ),\n",
    "]\n",
    "\n",
    "\n",
    "# PLOT AND EXPORT A LINEAR VIEW OF THE CONSTRUCT\n",
    "record = GraphicRecord(sequence_length=1000, features=features)\n",
    "ax, _ = record.plot(figure_width=5)\n",
    "ax.figure.savefig(\"graphic_record_defined_by_hand.png\")\n",
    "\n",
    "\n",
    "# PLOT AND EXPORT A CIRCULAR VIEW OF THE CONSTRUCT\n",
    "circular_rec = CircularGraphicRecord(sequence_length=1000, features=features)\n",
    "ax2, _ = circular_rec.plot(figure_width=4)\n",
    "ax2.figure.tight_layout()\n",
    "ax2.figure.savefig(\n",
    "    \"graphic_record_defined_by_hand_circular.png\", bbox_inches=\"tight\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load locally_highlighted_record.py\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "\n",
    "start, end = 1300, 2700\n",
    "\n",
    "\n",
    "def feature_properties(f):\n",
    "    \"\"\"Fade away all features not overlapping with [start, end]\"\"\"\n",
    "    if f.location.end < start or f.location.start > end:\n",
    "        return dict(color=\"white\", linecolor=\"grey\", label=None)\n",
    "    return {}\n",
    "\n",
    "\n",
    "translator = BiopythonTranslator(features_properties=feature_properties)\n",
    "graphic_record = translator.translate_record(\"example_sequence.gb\")\n",
    "ax, _ = graphic_record.plot(figure_width=12, elevate_outline_annotations=True)\n",
    "ax.fill_between(\n",
    "    [start, end], -10, 10, facecolor=\"peachpuff\", alpha=0.2, zorder=-1\n",
    ")\n",
    "\n",
    "ax.figure.savefig('locally_highlighted_record.png', bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load multiline_plot.py\n",
    "\"\"\"In this example we plot a record fragment with sequence over multiple lines.\n",
    "\"\"\"\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "\n",
    "translator = BiopythonTranslator()\n",
    "graphic_record = translator.translate_record(\"example_sequence.gb\")\n",
    "subrecord = graphic_record.crop((1700, 2000))\n",
    "fig, axes = subrecord.plot_on_multiple_lines(\n",
    "    nucl_per_line=70, plot_sequence=True\n",
    ")\n",
    "fig.savefig(\"multiline_plot.svg\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load multipage_plot.py\n",
    "\"\"\"In this example we plot a record fragment with sequence over multiple lines.\n",
    "\"\"\"\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "\n",
    "\n",
    "class CustomTranslator(BiopythonTranslator):\n",
    "    def compute_feature_color(self, feature):\n",
    "        return {\n",
    "            \"restriction_site\": \"yellow\",\n",
    "            \"CDS\": \"orange\",\n",
    "            \"promoter\": \"darkblue\",\n",
    "            \"terminator\": \"lightblue\",\n",
    "        }[feature.type]\n",
    "\n",
    "\n",
    "translator = CustomTranslator()\n",
    "graphic_record = translator.translate_record(\"example_sequence.gb\")\n",
    "subrecord = graphic_record.crop((1800, 2750))\n",
    "subrecord.plot_on_multiple_pages(\n",
    "    \"multipage_plot.pdf\",\n",
    "    nucl_per_line=70,\n",
    "    lines_per_page=7,\n",
    "    plot_sequence=True,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load plot_with_bokeh.py\n",
    "\"\"\"Simple example with Bokeh output. Requires the Bokeh library installed.\n",
    "\"\"\"\n",
    "\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "from bokeh.resources import CDN\n",
    "from bokeh.embed import file_html\n",
    "\n",
    "record = BiopythonTranslator().translate_record(record=\"example_sequence.gb\")\n",
    "plot = record.plot_with_bokeh(figure_width=8)\n",
    "\n",
    "with open(\"plot_with_bokeh.html\", \"w+\") as f:\n",
    "    f.write(file_html(plot, CDN, \"Example Sequence\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load translator_with_cutom_colors.py\n",
    "\"\"\"An example with GIF generation at the end. How cool is that!\n",
    "\n",
    "This example requires the Moviepy library  installed (pip install moviepy).\n",
    "\n",
    "\"\"\"\n",
    "from Bio import Entrez, SeqIO\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "\n",
    "# DOWNLOAD THE PLASMID's RECORD FROM NCBI\n",
    "Entrez.email = \"zulko@egf.org\"\n",
    "handle = Entrez.efetch(\n",
    "    db=\"nucleotide\", id=1473096477, rettype=\"gb\", retmode=\"text\"\n",
    ")\n",
    "record = SeqIO.read(handle, \"genbank\")\n",
    "\n",
    "# CREATE THE GRAPHIC RECORD WITH DNA_FEATURES_VIEWER\n",
    "\n",
    "color_map = {\n",
    "    \"rep_origin\": \"yellow\",\n",
    "    \"CDS\": \"orange\",\n",
    "    \"regulatory\": \"red\",\n",
    "    \"misc_recomb\": \"darkblue\",\n",
    "    \"misc_feature\": \"lightblue\",\n",
    "}\n",
    "translator = BiopythonTranslator(\n",
    "    features_filters=(lambda f: f.type not in [\"gene\", \"source\"],),\n",
    "    features_properties=lambda f: {\"color\": color_map.get(f.type, \"white\")},\n",
    ")\n",
    "translator.max_line_length = 15\n",
    "graphic_record = translator.translate_record(record)\n",
    "ax, _ = graphic_record.plot(figure_width=8, strand_in_label_threshold=7)\n",
    "ax.figure.savefig(\"translator_with_custom_colors.png\", bbox_inches=\"tight\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load with_gc_plot.py\n",
    "\"\"\"In this example we plot a record's annotations on top of the curve of the\n",
    "local GC content in the record's sequence.\n",
    "\"\"\"\n",
    "import matplotlib.pyplot as plt\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def plot_local_gc_content(record, window_size, ax):\n",
    "    \"\"\"Plot windowed GC content on a designated Matplotlib ax.\"\"\"\n",
    "    def gc_content(s):\n",
    "        return 100.0 * len([c for c in s if c in \"GC\"]) / len(s)\n",
    "\n",
    "    yy = [\n",
    "        gc_content(record.seq[i : i + window_size])\n",
    "        for i in range(len(record.seq) - window_size)\n",
    "    ]\n",
    "    xx = np.arange(len(record.seq) - window_size) + 25\n",
    "    ax.fill_between(xx, yy, alpha=0.3)\n",
    "    ax.set_ylim(bottom=0)\n",
    "    ax.set_ylabel(\"GC(%)\")\n",
    "\n",
    "\n",
    "record = SeqIO.read(\"example_sequence.gb\", \"genbank\")\n",
    "translator = BiopythonTranslator()\n",
    "graphic_record = translator.translate_record(record)\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(\n",
    "    2, 1, figsize=(10, 3), sharex=True, gridspec_kw={\"height_ratios\": [4, 1]}\n",
    ")\n",
    "graphic_record.plot(ax=ax1, with_ruler=False, strand_in_label_threshold=4)\n",
    "plot_local_gc_content(record, window_size=50, ax=ax2)\n",
    "\n",
    "fig.tight_layout()  # Resize the figure to the right height\n",
    "fig.savefig(\"with_gc_plot.png\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dna_features_viewer import BiopythonTranslator\n",
    "\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "sequences = []\n",
    "\n",
    "Entrez.email = \"richard.squires@nih.gov\"\n",
    "identifiers = [\"U94848.1\"]\n",
    "\n",
    "with Entrez.efetch( db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=identifiers) as handle:\n",
    "    for seq_record in SeqIO.parse(handle, \"gb\"):\n",
    "        sequences.append(seq_record)\n",
    "\n",
    "count = SeqIO.write(sequences, \"U94848.1.gb\", \"gb\")\n",
    "print(\"Save {} sequences to a file\".format(count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MyCustomTranslator(BiopythonTranslator):\n",
    "    \"\"\"Custom translator implementing the following theme:\n",
    "\n",
    "    - Color terminators in green, CDS in blue, all other features in gold.\n",
    "    - Do not display features that are restriction sites unless they are BamHI\n",
    "    - Do not display labels for restriction sites.\n",
    "    - For CDS labels just write \"CDS here\" instead of the name of the gene.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def compute_feature_color(self, feature):\n",
    "        if feature.type == \"gene\":\n",
    "            return \"blue\"\n",
    "        elif feature.type == \"terminator\":\n",
    "            return \"green\"\n",
    "        else:\n",
    "            return \"gold\"\n",
    "\n",
    "    def compute_feature_label(self, feature):\n",
    "        if feature.type == 'restriction_site':\n",
    "            return None\n",
    "        elif feature.type == \"CDS\":\n",
    "            return \"CDS\"\n",
    "        else:\n",
    "            return BiopythonTranslator.compute_feature_label(self, feature)\n",
    "\n",
    "    def compute_filtered_features(self, features):\n",
    "        \"\"\"Do not display promoters. Just because.\"\"\"\n",
    "        return [\n",
    "            feature for feature in features\n",
    "            if (feature.type != \"restriction_site\")\n",
    "            or (\"BamHI\" in str(feature.qualifiers.get(\"label\", '')))\n",
    "        ]\n",
    "\n",
    "\n",
    "graphic_record = MyCustomTranslator().translate_record(\"U94848.1.gb\")\n",
    "ax, _ = graphic_record.plot(figure_width=100)\n",
    "ax.figure.tight_layout()\n",
    "ax.figure.savefig(\"U94848.1.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"In this example we plot a record's annotations on top of the curve of the\n",
    "local GC content in the record's sequence.\n",
    "\"\"\"\n",
    "import matplotlib.pyplot as plt\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def plot_local_gc_content(record, window_size, ax):\n",
    "    \"\"\"Plot windowed GC content on a designated Matplotlib ax.\"\"\"\n",
    "    def gc_content(s):\n",
    "        return 100.0 * len([c for c in s if c in \"GC\"]) / len(s)\n",
    "\n",
    "    yy = [\n",
    "        gc_content(record.seq[i : i + window_size])\n",
    "        for i in range(len(record.seq) - window_size)\n",
    "    ]\n",
    "    xx = np.arange(len(record.seq) - window_size) + 25\n",
    "    ax.fill_between(xx, yy, alpha=0.3)\n",
    "    ax.set_ylim(bottom=0)\n",
    "    ax.set_ylabel(\"GC(%)\")\n",
    "\n",
    "\n",
    "record = SeqIO.read(\"U94848.1.gb\", \"gb\")\n",
    "translator = BiopythonTranslator()\n",
    "graphic_record = translator.translate_record(record)\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(\n",
    "    2, 1, figsize=(30, 3), sharex=True, gridspec_kw={\"height_ratios\": [4, 1]}\n",
    ")\n",
    "graphic_record.plot(ax=ax1, with_ruler=False, strand_in_label_threshold=4)\n",
    "plot_local_gc_content(record, window_size=1000, ax=ax2)\n",
    "\n",
    "fig.tight_layout()  # Resize the figure to the right height\n",
    "fig.savefig(\"with_gc_plot.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"An example with GIF generation at the end. How cool is that!\n",
    "\n",
    "This example requires the Moviepy library  installed (pip install moviepy).\n",
    "\n",
    "\"\"\"\n",
    "from Bio import Entrez, SeqIO\n",
    "import moviepy.editor as mpe\n",
    "from moviepy.video.io.bindings import mplfig_to_npimage\n",
    "import matplotlib.pyplot as plt\n",
    "from dna_features_viewer import BiopythonTranslator, CircularGraphicRecord\n",
    "\n",
    "# DOWNLOAD THE PLASMID's RECORD FROM NCBI\n",
    "\n",
    "handle = Entrez.efetch(\n",
    "    db=\"nucleotide\", id=1473096477, rettype=\"gb\", retmode=\"text\"\n",
    ")\n",
    "record = SeqIO.read(handle, \"genbank\")\n",
    "\n",
    "# CREATE THE GRAPHIC RECORD WITH DNA_FEATURES_VIEWER\n",
    "\n",
    "color_map = {\n",
    "    \"rep_origin\": \"yellow\",\n",
    "    \"CDS\": \"orange\",\n",
    "    \"regulatory\": \"red\",\n",
    "    \"misc_recomb\": \"darkblue\",\n",
    "    \"misc_feature\": \"lightblue\",\n",
    "}\n",
    "translator = BiopythonTranslator(\n",
    "    features_filters=(lambda f: f.type not in [\"gene\", \"source\"],),\n",
    "    features_properties=lambda f: {\"color\": color_map.get(f.type, \"white\")},\n",
    ")\n",
    "translator.max_line_length = 15\n",
    "graphic_record = translator.translate_record(\n",
    "    record, record_class=CircularGraphicRecord\n",
    ")\n",
    "graphic_record.labels_spacing = 15\n",
    "\n",
    "# ANIMATE INTO A GIF WITH MOVIEPY\n",
    "\n",
    "duration = 5\n",
    "\n",
    "\n",
    "def make_frame(t):\n",
    "    top_nucleotide_index = t * graphic_record.sequence_length / duration\n",
    "    graphic_record.top_position = top_nucleotide_index\n",
    "    ax, _ = graphic_record.plot(figure_width=8, figure_height=11)\n",
    "    ax.set_ylim(top=2)\n",
    "    np_image = mplfig_to_npimage(ax.figure)\n",
    "    plt.close(ax.figure)\n",
    "    return np_image\n",
    "\n",
    "\n",
    "clip = mpe.VideoClip(make_frame, duration=duration)\n",
    "small_clip = clip.crop(x1=60, x2=-60, y1=100, y2=-100).resize(0.5)\n",
    "small_clip.write_gif(\"example_with_gif.gif\", fps=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Simple example where a few features are defined \"by hand\" and are displayed\n",
    "and exported as PNG, first with a linear view, then with a circular\n",
    "view.\n",
    "\"\"\"\n",
    "\n",
    "from dna_features_viewer import (\n",
    "    GraphicFeature,\n",
    "    GraphicRecord,\n",
    "    CircularGraphicRecord,\n",
    ")\n",
    "\n",
    "features = [\n",
    "    GraphicFeature(\n",
    "        start=5, end=20, strand=+1, color=\"#ffd700\", label=\"Small feature\"\n",
    "    ),\n",
    "    GraphicFeature(\n",
    "        start=20,\n",
    "        end=500,\n",
    "        strand=+1,\n",
    "        color=\"#ffcccc\",\n",
    "        label=\"Gene 1 with a very long name\",\n",
    "    ),\n",
    "    GraphicFeature(\n",
    "        start=400, end=700, strand=-1, color=\"#cffccc\", label=\"Gene 2\"\n",
    "    ),\n",
    "    GraphicFeature(\n",
    "        start=600, end=900, strand=+1, color=\"#ccccff\", label=\"Gene 3\"\n",
    "    ),\n",
    "]\n",
    "\n",
    "\n",
    "# PLOT AND EXPORT A LINEAR VIEW OF THE CONSTRUCT\n",
    "record = GraphicRecord(sequence_length=1000, features=features)\n",
    "ax, _ = record.plot(figure_width=5)\n",
    "ax.figure.savefig(\"graphic_record_defined_by_hand.png\")\n",
    "\n",
    "\n",
    "# PLOT AND EXPORT A CIRCULAR VIEW OF THE CONSTRUCT\n",
    "circular_rec = CircularGraphicRecord(sequence_length=1000, features=features)\n",
    "ax2, _ = circular_rec.plot(figure_width=4)\n",
    "ax2.figure.tight_layout()\n",
    "ax2.figure.savefig(\n",
    "    \"graphic_record_defined_by_hand_circular.png\", bbox_inches=\"tight\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load example_with_gif.py\n",
    "\"\"\"An example with GIF generation at the end. How cool is that!\n",
    "\n",
    "This example requires the Moviepy library  installed (pip install moviepy).\n",
    "\n",
    "\"\"\"\n",
    "from Bio import Entrez, SeqIO\n",
    "import moviepy.editor as mpe\n",
    "from moviepy.video.io.bindings import mplfig_to_npimage\n",
    "import matplotlib.pyplot as plt\n",
    "from dna_features_viewer import BiopythonTranslator, CircularGraphicRecord\n",
    "\n",
    "# DOWNLOAD THE PLASMID's RECORD FROM NCBI\n",
    "\n",
    "handle = Entrez.efetch(\n",
    "    db=\"nucleotide\", id=1473096477, rettype=\"gb\", retmode=\"text\"\n",
    ")\n",
    "record = SeqIO.read(handle, \"genbank\")\n",
    "\n",
    "# CREATE THE GRAPHIC RECORD WITH DNA_FEATURES_VIEWER\n",
    "\n",
    "color_map = {\n",
    "    \"rep_origin\": \"yellow\",\n",
    "    \"CDS\": \"orange\",\n",
    "    \"regulatory\": \"red\",\n",
    "    \"misc_recomb\": \"darkblue\",\n",
    "    \"misc_feature\": \"lightblue\",\n",
    "}\n",
    "translator = BiopythonTranslator(\n",
    "    features_filters=(lambda f: f.type not in [\"gene\", \"source\"],),\n",
    "    features_properties=lambda f: {\"color\": color_map.get(f.type, \"white\")},\n",
    ")\n",
    "translator.max_line_length = 15\n",
    "graphic_record = translator.translate_record(\n",
    "    record, record_class=CircularGraphicRecord\n",
    ")\n",
    "graphic_record.labels_spacing = 15\n",
    "\n",
    "# ANIMATE INTO A GIF WITH MOVIEPY\n",
    "\n",
    "duration = 5\n",
    "\n",
    "\n",
    "def make_frame(t):\n",
    "    top_nucleotide_index = t * graphic_record.sequence_length / duration\n",
    "    graphic_record.top_position = top_nucleotide_index\n",
    "    ax, _ = graphic_record.plot(figure_width=8, figure_height=11)\n",
    "    ax.set_ylim(top=2)\n",
    "    np_image = mplfig_to_npimage(ax.figure)\n",
    "    plt.close(ax.figure)\n",
    "    return np_image\n",
    "\n",
    "\n",
    "clip = mpe.VideoClip(make_frame, duration=duration)\n",
    "small_clip = clip.crop(x1=60, x2=-60, y1=100, y2=-100).resize(0.5)\n",
    "small_clip.write_gif(\"example_with_gif.gif\", fps=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load cartoon_style.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
