{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Source of the materials**: Biopython cookbook (Adapted)\n",
    "<font color='red'>\n",
    "New status: Draft</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "BLAST\n",
    "===\n",
    "[Running BLAST over the Internet](#Running-BLAST-over-the-Internet)\n",
    "\n",
    "[Saving blast output](#Saving-blast-output)\n",
    "\n",
    "[Running BLAST locally](#Running-BLAST-locally)\n",
    "\n",
    "[Parsing BLAST output](#Parsing-BLAST-output)\n",
    "\n",
    "[The BLAST record class](#The-BLAST-record-class)\n",
    "\n",
    "[Parsing plain-text BLAST output](#Deprecated-BLAST-parsers)\n",
    "\n",
    "Hey, everybody loves BLAST right? I mean, geez, how can it get any\n",
    "easier to do comparisons between one of your sequences and every other\n",
    "sequence in the known world? But, of course, this section isn’t about\n",
    "how cool BLAST is, since we already know that. It is about the problem\n",
    "with BLAST – it can be really difficult to deal with the volume of data\n",
    "generated by large runs, and to automate BLAST runs in general.\n",
    "\n",
    "Fortunately, the Biopython folks know this only too well, so they’ve\n",
    "developed lots of tools for dealing with BLAST and making things much\n",
    "easier. This section details how to use these tools and do useful things\n",
    "with them.\n",
    "\n",
    "Dealing with BLAST can be split up into two steps, both of which can be\n",
    "done from within Biopython. Firstly, running BLAST for your query\n",
    "sequence(s), and getting some output. Secondly, parsing the BLAST output\n",
    "in Python for further analysis.\n",
    "\n",
    "Your first introduction to running BLAST was probably via the NCBI\n",
    "web-service. In fact, there are lots of ways you can run BLAST, which\n",
    "can be categorised in several ways. The most important distinction is\n",
    "running BLAST locally (on your own machine), and running BLAST remotely\n",
    "(on another machine, typically the NCBI servers). We’re going to start\n",
    "this chapter by invoking the NCBI online BLAST service from within a\n",
    "Python script.\n",
    "\n",
    "*NOTE*: The following Chapter \\[chapter:searchio\\] describes\n",
    "`Bio.SearchIO`, an *experimental* module in Biopython. We intend this to\n",
    "ultimately replace the older `Bio.Blast` module, as it provides a more\n",
    "general framework handling other related sequence searching tools as\n",
    "well. However, until that is declared stable, for production code please\n",
    "continue to use the `Bio.Blast` module for dealing with NCBI BLAST."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running BLAST over the Internet\n",
    "-----------------------------\n",
    "\n",
    "We use the function `qblast()` in the `Bio.Blast.NCBIWWW` module to call\n",
    "the online version of BLAST. This has three non-optional arguments:\n",
    "\n",
    "-   The first argument is the blast program to use for the search, as a\n",
    "    lower case string. The options and descriptions of the programs are\n",
    "    available at\n",
    "    <https://blast.ncbi.nlm.nih.gov/Blast.cgi.>. Currently\n",
    "    `qblast` only works with blastn, blastp, blastx, tblast and tblastx.\n",
    "\n",
    "-   The second argument specifies the databases to search against.\n",
    "    Again, the options for this are available on the NCBI web pages at\n",
    "    <http://www.ncbi.nlm.nih.gov/BLAST/blast_databases.shtml>.\n",
    "\n",
    "-   The third argument is a string containing your query sequence. This\n",
    "    can either be the sequence itself, the sequence in fasta format, or\n",
    "    an identifier like a GI number.\n",
    "\n",
    "The `qblast` function also take a number of other option arguments which\n",
    "are basically analogous to the different parameters you can set on the\n",
    "BLAST web page. We’ll just highlight a few of them here:\n",
    "\n",
    "-   The argument url_base sets the base URL for running BLAST over the internet. By default it connects to the NCBI, but one can use this to connect to an instance of NCBI BLAST running in the cloud. Please refer to the documentation for the qblast function for further details\n",
    "-   The `qblast` function can return the BLAST results in various\n",
    "    formats, which you can choose with the optional `format_type`\n",
    "    keyword: `\"HTML\"`, `\"Text\"`, `\"ASN.1\"`, or `\"XML\"`. The default is\n",
    "    `\"XML\"`, as that is the format expected by the parser, described in\n",
    "    section \\[sec:parsing-blast\\] below.\n",
    "\n",
    "-   The argument `expect` sets the expectation or e-value threshold.\n",
    "\n",
    "For more about the optional BLAST arguments, we refer you to the NCBI’s\n",
    "own documentation, or that built into Biopython:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "help(NCBIWWW.qblast)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Note that the default settings on the NCBI BLAST website are not quite\n",
    "the same as the defaults on QBLAST. If you get different results, you’ll\n",
    "need to check the parameters (e.g., the expectation value threshold and\n",
    "the gap values).\n",
    "\n",
    "For example, if you have a nucleotide sequence you want to search\n",
    "against the nucleotide database (nt) using BLASTN, and you know the GI\n",
    "number of your query sequence, you can use:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", \"8332116\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Alternatively, if we have our query sequence already in a FASTA\n",
    "formatted file, we just need to open the file and read in this record as\n",
    "a string, and use that as the query argument:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "fasta_string = open(\"data/m_cold.fasta\").read()\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", fasta_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "We could also have read in the FASTA file as a `SeqRecord` and then\n",
    "supplied just the sequence itself:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "from Bio import SeqIO\n",
    "record = SeqIO.read(\"data/m_cold.fasta\", format=\"fasta\")\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", record.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Supplying just the sequence means that BLAST will assign an identifier\n",
    "for your sequence automatically. You might prefer to use the `SeqRecord`\n",
    "object’s format method to make a FASTA string (which will include the\n",
    "existing identifier):\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "from Bio import SeqIO\n",
    "record = SeqIO.read(\"data/m_cold.fasta\", format=\"fasta\")\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", record.format(\"fasta\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This approach makes more sense if you have your sequence(s) in a\n",
    "non-FASTA file format which you can extract using `Bio.SeqIO` (see\n",
    "Chapter [5 - Sequence Input and Output](05 - Sequence Input and Output.ipynb).)\n",
    "\n",
    "Whatever arguments you give the `qblast()` function, you should get back\n",
    "your results in a handle object (by default in XML format). The next\n",
    "step would be to parse the XML output into Python objects representing\n",
    "the search results (Section \\[sec:parsing-blast\\]), but you might want\n",
    "to save a local copy of the output file first. I find this especially\n",
    "useful when debugging my code that extracts info from the BLAST results\n",
    "(because re-running the online search is slow and wastes the NCBI\n",
    "computer time)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Saving blast output\n",
    "------------------\n",
    "\n",
    "We need to be a bit careful since we can use `result_handle.read()` to\n",
    "read the BLAST output only once – calling `result_handle.read()` again\n",
    "returns an empty string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/my_blast.xml\", \"w\") as out_handle:\n",
    "    out_handle.write(result_handle.read())\n",
    "    result_handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "After doing this, the results are in the file `my_blast.xml` and the\n",
    "original handle has had all its data extracted (so we closed it).\n",
    "However, the `parse` function of the BLAST parser (described\n",
    "in \\[sec:parsing-blast\\]) takes a file-handle-like object, so we can\n",
    "just open the saved file for input:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " result_handle = open(\"data/my_blast.xml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we’ve got the BLAST results back into a handle again, we are\n",
    "ready to do something with them, so this leads us right into the parsing\n",
    "section (see Section \\[sec:parsing-blast\\] below). You may want to jump\n",
    "ahead to that now …."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running BLAST locally\n",
    "--------------------\n",
    "\n",
    "### Introduction\n",
    "\n",
    "Running BLAST locally (as opposed to over the internet, see\n",
    "Section \\[sec:running-www-blast\\]) has at least major two advantages:\n",
    "\n",
    "-   Local BLAST may be faster than BLAST over the internet;\n",
    "\n",
    "-   Local BLAST allows you to make your own database to search for\n",
    "    sequences against.\n",
    "\n",
    "Dealing with proprietary or unpublished sequence data can be another\n",
    "reason to run BLAST locally. You may not be allowed to redistribute the\n",
    "sequences, so submitting them to the NCBI as a BLAST query would not be\n",
    "an option.\n",
    "\n",
    "Unfortunately, there are some major drawbacks too – installing all the\n",
    "bits and getting it setup right takes some effort:\n",
    "\n",
    "-   Local BLAST requires command line tools to be installed.\n",
    "\n",
    "-   Local BLAST requires (large) BLAST databases to be setup (and\n",
    "    potentially kept up to date).\n",
    "\n",
    "To further confuse matters there are several different BLAST packages\n",
    "available, and there are also other tools which can produce imitation\n",
    "BLAST output files, such as BLAT.\n",
    "\n",
    "### Standalone NCBI BLAST+\n",
    "\n",
    "The “new” [NCBI\n",
    "BLAST+](http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download)\n",
    "suite was released in 2009. This replaces the old NCBI “legacy” BLAST\n",
    "package (see below).\n",
    "\n",
    "This section will show briefly how to use these tools from within\n",
    "Python. If you have already read or tried the alignment tool examples in\n",
    "Section \\[sec:alignment-tools\\] this should all seem quite\n",
    "straightforward. First, we construct a command line string (as you would\n",
    "type in at the command line prompt if running standalone BLAST by hand).\n",
    "Then we can execute this command from within Python.\n",
    "\n",
    "For example, taking a FASTA file of gene nucleotide sequences, you might\n",
    "want to run a BLASTX (translation) search against the non-redundant (NR)\n",
    "protein database. Assuming you (or your systems administrator) has\n",
    "downloaded and installed the NR database, you might run:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "blastx -query opuntia.fasta -db nr -out opuntia.xml -evalue 0.001 -outfmt 5\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "This should run BLASTX against the NR database, using an expectation\n",
    "cut-off value of $0.001$ and produce XML output to the specified file\n",
    "(which we can then parse). On my computer this takes about six minutes -\n",
    "a good reason to save the output to a file so you can repeat any\n",
    "analysis as needed.\n",
    "\n",
    "From within Biopython we can use the NCBI BLASTX wrapper from the\n",
    "`Bio.Blast.Applications` module to build the command line string, and\n",
    "run it:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast.Applications import NcbiblastxCommandline\n",
    "help(NcbiblastxCommandline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "blastx_cline = NcbiblastxCommandline(query=\"opuntia.fasta\", db=\"nr\", evalue=0.001,\n",
    "outfmt=5, out=\"opuntia.xml\")\n",
    "blastx_cline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(blastx_cline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# stdout, stderr = blastx_cline()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example there shouldn’t be any output from BLASTX to the\n",
    "terminal, so stdout and stderr should be empty. You may want to check\n",
    "the output file `opuntia.xml` has been created.\n",
    "\n",
    "As you may recall from earlier examples in the tutorial, the\n",
    "`opuntia.fasta` contains seven sequences, so the BLAST XML output should\n",
    "contain multiple results. Therefore use `Bio.Blast.NCBIXML.parse()` to\n",
    "parse it as described below in Section \\[sec:parsing-blast\\].\n",
    "\n",
    "### Other versions of BLAST\n",
    "\n",
    "NCBI BLAST+ (written in C++) was first released in 2009 as a replacement\n",
    "for the original NCBI “legacy” BLAST (written in C) which is no longer\n",
    "being updated. There were a lot of changes – the old version had a\n",
    "single core command line tool `blastall` which covered multiple\n",
    "different BLAST search types (which are now separate commands in\n",
    "BLAST+), and all the command line options were renamed. Biopython’s\n",
    "wrappers for the NCBI “legacy” BLAST tools have been deprecated and will\n",
    "be removed in a future release. To try to avoid confusion, we do not\n",
    "cover calling these old tools from Biopython in this tutorial.\n",
    "\n",
    "You may also come across [Washington University\n",
    "BLAST](http://blast.wustl.edu/) (WU-BLAST), and its successor, [Advanced\n",
    "Biocomputing BLAST](http://blast.advbiocomp.com) (AB-BLAST, released in\n",
    "2009, not free/open source). These packages include the command line\n",
    "tools `wu-blastall` and `ab-blastall`, which mimicked `blastall` from\n",
    "the NCBI “legacy” BLAST suite. Biopython does not currently provide\n",
    "wrappers for calling these tools, but should be able to parse any NCBI\n",
    "compatible output from them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parsing BLAST output\n",
    "-------------------\n",
    "\n",
    "As mentioned above, BLAST can generate output in various formats, such\n",
    "as XML, HTML, and plain text. Originally, Biopython had parsers for\n",
    "BLAST plain text and HTML output, as these were the only output formats\n",
    "offered at the time. Unfortunately, the BLAST output in these formats\n",
    "kept changing, each time breaking the Biopython parsers. Our HTML BLAST\n",
    "parser has been removed, but the plain text BLAST parser is still\n",
    "available (see Section \\[sec:parsing-blast-deprecated\\]). Use it at your\n",
    "own risk, it may or may not work, depending on which BLAST version\n",
    "you’re using.\n",
    "\n",
    "As keeping up with changes in BLAST became a hopeless endeavor,\n",
    "especially with users running different BLAST versions, we now recommend\n",
    "to parse the output in XML format, which can be generated by recent\n",
    "versions of BLAST. Not only is the XML output more stable than the plain\n",
    "text and HTML output, it is also much easier to parse automatically,\n",
    "making Biopython a whole lot more stable.\n",
    "\n",
    "You can get BLAST output in XML format in various ways. For the parser,\n",
    "it doesn’t matter how the output was generated, as long as it is in the\n",
    "XML format.\n",
    "\n",
    "-   You can use Biopython to run BLAST over the internet, as described\n",
    "    in section \\[sec:running-www-blast\\].\n",
    "\n",
    "-   You can use Biopython to run BLAST locally, as described\n",
    "    in section \\[sec:running-local-blast\\].\n",
    "\n",
    "-   You can do the BLAST search yourself on the NCBI site through your\n",
    "    web browser, and then save the results. You need to choose XML as\n",
    "    the format in which to receive the results, and save the final BLAST\n",
    "    page you get (you know, the one with all of the\n",
    "    interesting results!) to a file.\n",
    "\n",
    "-   You can also run BLAST locally without using Biopython, and save the\n",
    "    output in a file. Again, you need to choose XML as the format in\n",
    "    which to receive the results.\n",
    "\n",
    "The important point is that you do not have to use Biopython scripts to\n",
    "fetch the data in order to be able to parse it. Doing things in one of\n",
    "these ways, you then need to get a handle to the results. In Python, a\n",
    "handle is just a nice general way of describing input to any info source\n",
    "so that the info can be retrieved using `read()` and `readline()`\n",
    "functions (see Section <span>sec:appendix-handles</span>).\n",
    "\n",
    "If you followed the code above for interacting with BLAST through a\n",
    "script, then you already have `result_handle`, the handle to the BLAST\n",
    "results. For example, using a GI number to do an online search:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", \"8332116\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "If instead you ran BLAST some other way, and have the BLAST output (in\n",
    "XML format) in the file `my_blast.xml`, all you need to do is to open\n",
    "the file for reading:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_handle = open(\"data/my_blast.xml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Now that we’ve got a handle, we are ready to parse the output. The code\n",
    "to parse it is really quite small. If you expect a single BLAST result\n",
    "(i.e., you used a single query):\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "blast_record = NCBIXML.read(result_handle)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "or, if you have lots of results (i.e., multiple query sequences):\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "blast_records = NCBIXML.parse(result_handle)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Just like `Bio.SeqIO` and `Bio.AlignIO` (see\n",
    "Chapters \\[chapter:Bio.SeqIO\\] and \\[chapter:Bio.AlignIO\\]), we have a\n",
    "pair of input functions, `read` and `parse`, where `read` is for when\n",
    "you have exactly one object, and `parse` is an iterator for when you can\n",
    "have lots of objects – but instead of getting `SeqRecord` or\n",
    "`MultipleSeqAlignment` objects, we get BLAST record objects.\n",
    "\n",
    "To be able to handle the situation where the BLAST file may be huge,\n",
    "containing thousands of results, `NCBIXML.parse()` returns an iterator.\n",
    "In plain English, an iterator allows you to step through the BLAST\n",
    "output, retrieving BLAST records one by one for each BLAST search\n",
    "result:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "blast_records = NCBIXML.parse(result_handle)\n",
    "blast_record = next(blast_records)\n",
    "print(blast_record.database_sequences)\n",
    "# # ... do something with blast_record"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Or, you can use a `for`-loop. Note though that you can step through the BLAST records only once. Usually, from each BLAST record you would save the information that you are interested in. If you want to save all returned BLAST records, you can convert the iterator into a list:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for blast_record in blast_records:\n",
    "#Do something with blast_records\n",
    "    blast_records = list(blast_records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_records = list(blast_records)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now you can access each BLAST record in the list with an index as usual. If your BLAST file is huge though, you may run into memory problems trying to save them all in a list.\n",
    "\n",
    "Usually, you’ll be running one BLAST search at a time. Then, all you need to do is to pick up the first (and only) BLAST record in blast_records:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "blast_records = NCBIXML.parse(result_handle)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "I guess by now you’re wondering what is in a BLAST record.\n",
    "\n",
    "The BLAST record class\n",
    "---------------------\n",
    "\n",
    "A BLAST Record contains everything you might ever want to extract from\n",
    "the BLAST output. Right now we’ll just show an example of how to get\n",
    "some info out of the BLAST report, but if you want something in\n",
    "particular that is not described here, look at the info on the record\n",
    "class in detail, and take a gander into the code or automatically\n",
    "generated documentation – the docstrings have lots of good info about\n",
    "what is stored in each piece of information.\n",
    "\n",
    "To continue with our example, let’s just print out some summary info\n",
    "about all hits in our blast report greater than a particular threshold.\n",
    "The following code does this:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "E_VALUE_THRESH = 0.04"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "outputPrepend"
    ]
   },
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "result_handle = open(\"data/my_blast.xml\", \"r\")\n",
    "blast_records = NCBIXML.parse(result_handle)\n",
    "\n",
    "for alignment in blast_record.alignments:\n",
    "    for hsp in alignment.hsps:\n",
    "        if hsp.expect < E_VALUE_THRESH:\n",
    "            print(\"****Alignment****\")\n",
    "            print(\"sequence:\", alignment.title)\n",
    "            print(\"length:\", alignment.length)\n",
    "            print(\"e value:\", hsp.expect)\n",
    "            print(hsp.query[0:75] + \"...\")\n",
    "            print(hsp.match[0:75] + \"...\")\n",
    "            print(hsp.sbjct[0:75] + \"...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will print out summary reports like the following: ****Alignment**** sequence: >gb|AF283004.1|AF283004 Arabidopsis thaliana cold acclimation protein WCOR413-like protein alpha form mRNA, complete cds\n",
    "length: 783 \n",
    "e value: 0.034 \n",
    "tacttgttgatattggatcgaacaaactggagaaccaacatgctcacgtcacttttagtcccttacatattcctc...\n",
    "||||||||| | ||||||||||| || |||| || || |||||||| |||||| | | |||||||| ||| | |...\n",
    "tacttgttggtgttggatcgaaccaattggaagacgaatatgctcacatcacttctcattccttacatcttcttc...\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Basically, you can do anything you want to with the info in the BLAST\n",
    "report once you have parsed it. This will, of course, depend on what you\n",
    "want to use it for, but hopefully this helps you get started on doing\n",
    "what you need to do!\n",
    "\n",
    "An important consideration for extracting information from a BLAST\n",
    "report is the type of objects that the information is stored in. In\n",
    "Biopython, the parsers return `Record` objects, either `Blast` or\n",
    "`PSIBlast` depending on what you are parsing. These objects are defined\n",
    "in `Bio.Blast.Record` and are quite complete.\n",
    "\n",
    "Here are my attempts at UML class diagrams for the `Blast` and\n",
    "`PSIBlast` record classes. If you are good at UML and see\n",
    "mistakes/improvements that can be made, please let me know. The Blast\n",
    "class diagram is shown in the next figure\n",
    "\n",
    "![Class diagram for the Blast Record class representing all of the info\n",
    "in a BLAST report<span\n",
    "data-label=\"fig:blastrecord\"></span>](images/BlastRecord.png)\n",
    "\n",
    "The PSIBlast record object is similar, but has support for the rounds\n",
    "that are used in the iteration steps of PSIBlast. The class diagram for\n",
    "PSIBlast is shown in the next figure.\n",
    "\n",
    "![Class diagram for the PSIBlast Record class.<span\n",
    "data-label=\"fig:psiblastrecord\"></span>](images/PSIBlastRecord.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dealing with PSI-BLAST\n",
    "---------------------\n",
    "You can run the standalone version of PSI-BLAST (the legacy NCBI command line tool blastpgp, or its replacement psiblast) using the wrappers in Bio.Blast.Applications module. At the time of writing, the NCBI does not appear to support tools running a PSI-BLAST search via the internet. Note that the Bio.Blast.NCBIXML parser can read the XML output from current versions of PSI-BLAST, but information like which sequences in each iteration is new or reused isn’t present in the XML file. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dealing with RPS-BLAST\n",
    "---------------------\n",
    "You can run the standalone version of RPS-BLAST (either the legacy NCBI command line tool rpsblast, or its replacement with the same name) using the wrappers in Bio.Blast.Applications module. At the time of writing, the NCBI does not appear to support tools running an RPS-BLAST search via the internet. You can use the Bio.Blast.NCBIXML parser to read the XML output from current versions of RPSBLAST.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3-final"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
