{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "KEGG\n",
    "====\n",
    "\n",
    "KEGG (<http://www.kegg.jp/>) is a database resource for understanding\n",
    "high-level functions and utilities of the biological system, such as the\n",
    "cell, the organism and the ecosystem, from molecular-level information,\n",
    "especially large-scale molecular datasets generated by genome sequencing\n",
    "and other high-throughput experimental technologies.\n",
    "\n",
    "Please note that the KEGG parser implementation in Biopython is\n",
    "incomplete. While the KEGG website indicates many flat file formats,\n",
    "only parsers and writers for compound, enzyme, and map are currently\n",
    "implemented. However, a generic parser is implemented to handle the\n",
    "other formats.\n",
    "\n",
    "Parsing KEGG records\n",
    "--------------------\n",
    "\n",
    "Parsing a KEGG record is as simple as using any other file format parser\n",
    "in Biopython. (Before running the following codes, please open\n",
    "http://rest.kegg.jp/get/ec:5.4.2.2 with your web browser and save it as\n",
    "ec\\_5.4.2.2.txt.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget http://rest.kegg.jp/get/ec:5.4.2.2 -O ec_5.4.2.2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.KEGG import Enzyme\n",
    "\n",
    "records = Enzyme.parse(open(\"ec_5.4.2.2.txt\"))\n",
    "\n",
    "record = list(records)[0]\n",
    "\n",
    "record.classname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "record.entry"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following section will shows how to download the above enzyme using\n",
    "the KEGG api as well as how to use the generic parser with data that\n",
    "does not have a custom parser implemented.\n",
    "\n",
    "Querying the KEGG API\n",
    "---------------------\n",
    "\n",
    "Biopython has full support for the querying of the KEGG api. Querying\n",
    "all KEGG endpoints are supported; all methods documented by KEGG\n",
    "(<http://www.kegg.jp/kegg/rest/keggapi.html>) are supported. The\n",
    "interface has some validation of queries which follow rules defined on\n",
    "the KEGG site. However, invalid queries which return a 400 or 404 must\n",
    "be handled by the user.\n",
    "\n",
    "First, here is how to extend the above example by downloading the\n",
    "relevant enzyme and passing it through the Enzyme parser."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.KEGG import REST\n",
    "\n",
    "from Bio.KEGG import Enzyme\n",
    "\n",
    "request = REST.kegg_get(\"ec:5.4.2.2\")\n",
    "\n",
    "open(\"ec_5.4.2.2.txt\", 'w').write(request.read().decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = Enzyme.parse(open(\"ec_5.4.2.2.txt\"))\n",
    "\n",
    "record = list(records)[0]\n",
    "\n",
    "record.classname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "record.entry"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, here’s a more realistic example which shows a combination of\n",
    "querying the KEGG API. This will demonstrate how to extract a unique set\n",
    "of all human pathway gene symbols which relate to DNA repair. The steps\n",
    "that need to be taken to do so are as follows. First, we need to get a\n",
    "list of all human pathways. Secondly, we need to filter those for ones\n",
    "which relate to “repair”. Lastly, we need to get a list of all the gene\n",
    "symbols in all repair pathways."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.KEGG import REST\n",
    "\n",
    "human_pathways = REST.kegg_list(\"pathway\", \"hsa\").read()\n",
    "\n",
    "human_pathways.decode(\"utf-8\").split(\"\\n\")[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter all human pathways for repair pathways\n",
    "repair_pathways = []\n",
    "for line in human_pathways.decode(\"utf-8\").rstrip().split(\"\\n\"):\n",
    "    entry, description = line.split(\"\\t\")\n",
    "    if \"repair\" in description:\n",
    "        repair_pathways.append(entry)\n",
    "\n",
    "repair_pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the genes for pathways and add them to a list\n",
    "repair_genes = [] \n",
    "for pathway in repair_pathways:\n",
    "    pathway_file = REST.kegg_get(pathway).read()  # query and read each pathway\n",
    "\n",
    "    # iterate through each KEGG pathway file, keeping track of which section\n",
    "    # of the file we're in, only read the gene in each pathway\n",
    "    current_section = None\n",
    "    for line in pathway_file.decode(\"utf-8\").rstrip().split(\"\\n\"):\n",
    "        section = line[:12].strip()  # section names are within 12 columns\n",
    "        if not section == \"\":\n",
    "            current_section = section\n",
    "        \n",
    "        if current_section == \"GENE\":\n",
    "            gene_identifiers, gene_description = line[12:].split(\"; \")\n",
    "            gene_id, gene_symbol = gene_identifiers.split()\n",
    "\n",
    "            if not gene_symbol in repair_genes:\n",
    "                repair_genes.append(gene_symbol)\n",
    "\n",
    "print(\"There are %d repair pathways and %d repair genes. The genes are:\" % \\\n",
    "        (len(repair_pathways), len(repair_genes)))\n",
    "print(\", \".join(repair_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The KEGG API wrapper is compatible with all endpoints. Usage is\n",
    "essentially replacing all slashes in the url with commas and using that\n",
    "list as arguments to the corresponding method in the KEGG module. Here\n",
    "are a few examples from the api documentation\n",
    "(<http://www.kegg.jp/kegg/docs/keggapi.html>).\n",
    "\n",
    "    /list/hsa:10458+ece:Z5100            -> REST.kegg_list([\"hsa:10458\", \"ece:Z5100\"])\n",
    "    /find/compound/300-310/mol_weight    -> REST.kegg_find(\"compound\", \"300-310\", \"mol_weight\")\n",
    "    /get/hsa:10458+ece:Z5100/aaseq      -> REST.kegg_get([\"hsa:10458\", \"ece:Z5100\"], \"aaseq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
